File: packages/python/src/llmwho/anthropic.py
"""Content-free normalization for direct Anthropic Messages API traffic."""

from __future__ import annotations

from typing import Any, Mapping, Optional
from urllib.parse import urlsplit


PROVIDER = "anthropic"
MESSAGES_OPERATION = "messages"


def is_messages_url(url: str) -> bool:
    """Return whether *url* is Anthropic's direct Messages endpoint."""

    try:
        parsed = urlsplit(url)
        return (
            (parsed.hostname or "").lower() == "api.anthropic.com"
            and parsed.path.rstrip("/") == "/v1/messages"
        )
    except Exception:
        return False


def _token(value: Any) -> Optional[int]:
    return value if type(value) is int and value >= 0 else None


def request_metadata(
    payload: Optional[Mapping[str, Any]], size: int
) -> tuple[dict[str, Any], Optional[str]]:
    """Normalize safe request metadata without retaining message content."""

    metadata: dict[str, Any] = {
        "operation": MESSAGES_OPERATION,
        "input_bytes": max(0, size),
    }
    claimed = None
    if payload:
        model = payload.get("model")
        if isinstance(model, str):
            claimed = model
            metadata["claimed_model"] = model
        stream = payload.get("stream")
        if isinstance(stream, bool):
            metadata["stream"] = stream
        messages = payload.get("messages")
        if isinstance(messages, list):
            metadata["role_count"] = len(messages)
    return metadata, claimed


def response_metadata(
    payload: Optional[Mapping[str, Any]], size: int, status_code: int
) -> tuple[dict[str, Any], Optional[str]]:
    """Normalize safe response metadata and Anthropic token accounting."""

    metadata: dict[str, Any] = {
        "status_code": status_code,
        "output_bytes": max(0, size),
    }
    declared = None
    if not payload:
        return metadata, declared

    model = payload.get("model")
    if isinstance(model, str):
        declared = model
        metadata["declared_model"] = model

    raw_usage = payload.get("usage")
    if isinstance(raw_usage, Mapping):
        input_tokens = _token(raw_usage.get("input_tokens"))
        output_tokens = _token(raw_usage.get("output_tokens"))
        usage: dict[str, int] = {}
        if input_tokens is not None:
            usage["input_tokens"] = input_tokens
        if output_tokens is not None:
            usage["output_tokens"] = output_tokens
        if input_tokens is not None and output_tokens is not None:
            usage["total_tokens"] = input_tokens + output_tokens
        if usage:
            metadata["usage"] = usage
    return metadata, declared

File: packages/node/src/anthropic.js
export const ANTHROPIC_PROVIDER = "anthropic";
export const MESSAGES_OPERATION = "messages";

export function isAnthropicMessagesUrl(value) {
  try {
    const url = new URL(value);
    return url.hostname.toLowerCase() === "api.anthropic.com"
      && url.pathname.replace(/\/+$/, "") === "/v1/messages";
  } catch {
    return false;
  }
}

function token(value) {
  return Number.isInteger(value) && value >= 0 ? value : undefined;
}

export function anthropicRequestMetadata(payload, bytes) {
  const metadata = {
    operation: MESSAGES_OPERATION,
    input_bytes: Math.max(0, bytes),
  };
  let requestedModel;
  if (payload && typeof payload === "object") {
    if (typeof payload.model === "string") {
      requestedModel = payload.model;
      metadata.requested_model = requestedModel;
    }
    if (typeof payload.stream === "boolean") metadata.stream = payload.stream;
    if (Array.isArray(payload.messages)) metadata.role_count = payload.messages.length;
  }
  return [metadata, requestedModel];
}

export function anthropicResponseMetadata(payload, bytes, status) {
  const metadata = {
    status_code: status,
    output_bytes: Math.max(0, bytes),
  };
  let declaredModel;
  if (!payload || typeof payload !== "object") return [metadata, declaredModel];

  if (typeof payload.model === "string") {
    declaredModel = payload.model;
  }
  if (payload.usage && typeof payload.usage === "object") {
    const inputTokens = token(payload.usage.input_tokens);
    const outputTokens = token(payload.usage.output_tokens);
    const usage = {};
    if (inputTokens !== undefined) usage.input_tokens = inputTokens;
    if (outputTokens !== undefined) usage.output_tokens = outputTokens;
    if (inputTokens !== undefined && outputTokens !== undefined) {
      usage.total_tokens = inputTokens + outputTokens;
    }
    if (Object.keys(usage).length) metadata.usage = usage;
  }
  return [metadata, declaredModel];
}
